platforms/allwinner-d1/src/dmac/mod.rs

```diff
@@ -13,167 +13,481 @@ use d1_pac::{
+                    // `DMAC_IRQ_EN1` register, instead.
+                    dmac.dmac_irq_en1
+                        .modify(|r, w| w.bits(set_queue_irq_en(idx - 8, r.bits())));
+                }
+            }
+        });
+        Self { _p: () }
     }
 
-    /// Allocates a new DMA [`Channel`].
-    pub fn allocate_channel(&self) -> Option<Channel> {
-        /// Set the `DMA_QUEUE_IRQ_EN` bit for the given channel index.
-        fn set_queue_irq_en(idx: u8, bits: u32) -> u32 {
-            bits | (1 << queue_irq_en_offset(idx))
+    /// Performs a DMA transfer described by the provided [`Descriptor`] on any
+    /// available free DMA [`Channel`].
+    ///
+    /// This function will first use [`Dmac::claim_channel`] to claim an
+    /// available DMA channel, waiting for one to become available if all
+    /// channels are currently in use. Once a channel has been acquired, it sets
+    /// the `src` and `dst` [`ChannelMode`]s using
+    /// [`Channel::set_channel_modes`]. Then, it performs the transfer using
+    /// [`Channel::transfer`].
+    ///
+    /// If multiple transfers will be made in sequence, it may be more efficient to
+    /// call [`Dmac::claim_channel`] once and perform multiple transfers on the same
+    /// channel, to avoid releasing and re-claiming a channel in between transfers.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the descriptor pointed to by `desc`, and the
+    /// associated memory region used by the transfer, is valid for as long as
+    /// the DMA transfer is active. When this function returns, the transfer has
+    /// completed, and it is safe to drop the descriptor. If this future is
+    /// cancelled, the transfer is cancelled and the descriptor
+    /// and its associated buffer may be dropped safely. However, **it is super
+    /// ultra not okay to [`core::mem::forget`] this future**. If you
+    /// `mem::forget` a DMA transfer future inside your driver, you deserve
+    /// whatever happens next.
+    ///
+    /// # Cancel Safety
+    ///
+    /// Dropping this future cancels the DMA transfer. If this future is
+    /// dropped, the descriptor and its associated memory region may also be
+    /// dropped safely.
+    ///
+    /// Of course, the transfer may still have completed partially, and if we
+    /// were writing to a device, the device may be unhappy to have only gotten
+    /// some of the data it wanted. Cancelling an incomplete transfer may result
+    /// in, for example, writing out half of a string to the UART, or only part
+    /// of a structured message over SPI, and so on. But, at least we don't have
+    /// abandoned DMA transfers running around in random parts of the heap you
+    /// probably wanted to use for normal stuff like having strings, or whatever
+    /// it is that people do on the computer.
+    ///
+    /// If the DMA transfer was a read rather than a write, cancelling a partial
```

Comment:
I'd say this isn't TOTALLY true, depending on how broad you are talking about, reads can still have side effects in external hardware.

Reply:
yeah, good point